clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -1,17 +1,210 @@
+
+          // We need to get the internal consumer as we have to intercept the control messages that we would normally
+          // filter out from the user
+          PubSubConsumerAdapter consumerAdapter = internalSeekConsumer.get().getPubSubConsumer();
+
+          Map<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> polledResults;
+          Map<Integer, Boolean> endOfPushConsumedPerPartitionMap = new HashMap<>();
+          Set<VeniceChangeCoordinate> checkpoints = new HashSet<>();
+
+          // Initialize map with all false entries for each partition
+          for (Integer partition: partitions) {
+            endOfPushConsumedPerPartitionMap.put(partition, false);
+          }
+
+          // poll until we get EOP for all partitions
+          while (true) {
+            polledResults = consumerAdapter.poll(5000L);
+            // Loop through all polled messages
+            for (Map.Entry<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> entry: polledResults
+                .entrySet()) {
+              PubSubTopicPartition pubSubTopicPartition = entry.getKey();
+              List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> messageList = entry.getValue();
+              for (PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> message: messageList) {
+                if (message.getKey().isControlMessage()) {
+                  ControlMessage controlMessage = (ControlMessage) message.getValue().getPayloadUnion();
+                  ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+                  if (controlMessageType.equals(ControlMessageType.END_OF_PUSH)) {
+                    // note down the partition and offset and mark that we've got the thing
+                    endOfPushConsumedPerPartitionMap.put(pubSubTopicPartition.getPartitionNumber(), true);
+                    VeniceChangeCoordinate coordinate = new VeniceChangeCoordinate(
+                        pubSubTopicPartition.getPubSubTopic().getName(),
+                        new ApacheKafkaOffsetPosition(message.getOffset()),
+                        pubSubTopicPartition.getPartitionNumber());
+                    checkpoints.add(coordinate);
+                    Set<Integer> unsubSet = new HashSet<>();
+                    unsubSet.add(pubSubTopicPartition.getPartitionNumber());
+                    internalSeekConsumer.get().unsubscribe(unsubSet);
+                    // No need to look at the rest of the messages for this partition that we might have polled
+                    break;
+                  }
+                }
+              }
+            }
+            if (endOfPushConsumedPerPartitionMap.values().stream().allMatch(e -> e)) {
+              // We polled all EOP messages, stop polling!
+              break;
+            }
+          }
+          this.seekToCheckpoint(checkpoints).get();
```

Comment:
Is it possible we will get the thread seek to EOP endlessly here, as `versionSwapDetectionThread` will keep triggering `seekToEndOfPush` and the main consumer always seek to checkpoints and then poll and then seekToCheckpoint?

Reply:
Not sure I follow, but I don't think so.  versionSwapDetectionThread is single threaded and we do a blocking call on seekToCheckpoint, and will only trigger the switch if the consumer isn't on the current version topic.  So I don't understand how this would loop.

Reply:
I think I get the idea. This switch will only be triggered with new current version detected by the thread. So it might be good to log here as we jumped to the newer version and the EOP postition we find for debugging purpose.